File: backend/app/schemas/donation.py
from __future__ import annotations

from datetime import date, datetime
from enum import StrEnum
from typing import Annotated, Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.core.sanitize import (
    OptionalFreeTextSanitizedStr,
    OptionalSanitizedStr,
    SanitizedStr,
)


class DonationPaymentMethod(StrEnum):
    UPI = "upi"
    QR_CODE = "qr_code"
    GPAY = "gpay"
    CASH = "cash"
    CHEQUE = "cheque"
    DEBIT_CARD = "debit_card"
    CREDIT_CARD = "credit_card"
    INTERNATIONAL_CARD = "international_card"
    STRIPE = "stripe"
    NETBANKING = "netbanking"
    ONLINE_BANKING = "online_banking"
    OTHER = "other"


UTR_REQUIRED_METHODS = frozenset(
    {
        DonationPaymentMethod.UPI,
        DonationPaymentMethod.QR_CODE,
        DonationPaymentMethod.GPAY,
        DonationPaymentMethod.CHEQUE,
    }
)

RECEIPT_FROZEN_FIELDS = frozenset(
    {
        "donor_name",
        "donor_email",
        "donor_phone",
        "donor_address",
        "donor_pan",
        "donor_type",
        "amount",
        "currency",
        "purpose",
        "payment_method",
        "transaction_reference",
    }
)

ANONYMOUS_DISPLAY_NAME = "Anonymous"


class DonationPaymentStatus(StrEnum):
    PENDING = "pending"
    PAID = "paid"
    FAILED = "failed"
    REFUNDED = "refunded"


class DonationPurpose(StrEnum):
    GENERAL = "general"
    EDUCATION = "education"
    HEALTHCARE = "healthcare"
    LIVELIHOOD = "livelihood"
    RELIEF = "relief"
    ENVIRONMENT = "environment"


class DonorType(StrEnum):
    INDIVIDUAL = "individual"
    CORPORATE = "corporate"
    TRUST = "trust"
    FOUNDATION = "foundation"
    OTHER = "other"


def _normalize_reference(value: str | None) -> str | None:
    if value is None:
        return None
    trimmed = value.strip()
    return trimmed or None


def _require_phone_unless_anonymous(
    *, is_anonymous: bool, donor_phone: str | None
) -> str | None:
    phone = _normalize_reference(donor_phone)
    if is_anonymous:
        if phone is not None and len(phone) < 7:
            raise ValueError("donor_phone must be at least 7 characters when provided")
        return phone
    if phone is None or len(phone) < 7:
        raise ValueError("donor_phone is required unless donation is anonymous")
    return phone


class DonorBase(BaseModel):
    donor_name: SanitizedStr = Field(..., min_length=2, max_length=140)
    donor_email: str | None = Field(default=None, pattern=r"^[^@\s]+@[^@\s]+\.[^@\s]+$")
    donor_phone: OptionalSanitizedStr = Field(default=None, max_length=20)
    donor_address: OptionalSanitizedStr = Field(default=None, max_length=260)
    donor_pan: Annotated[
        str | None, Field(default=None, pattern=r"^[A-Z]{5}[0-9]{4}[A-Z]$")
    ] = None
    donor_type: DonorType = DonorType.INDIVIDUAL


class DonationBase(DonorBase):
    amount: float = Field(..., gt=0)
    currency: str = Field(default="INR", min_length=3, max_length=3)
    purpose: DonationPurpose = DonationPurpose.GENERAL
    payment_method: DonationPaymentMethod = DonationPaymentMethod.UPI
    payment_status: DonationPaymentStatus = DonationPaymentStatus.PENDING
    transaction_reference: OptionalSanitizedStr = Field(default=None, max_length=120)
    donation_date: date = Field(default_factory=date.today)
    notes: OptionalFreeTextSanitizedStr = Field(default=None, max_length=500)
    gateway_provider: OptionalSanitizedStr = Field(default=None, max_length=40)
    gateway_session_id: OptionalSanitizedStr = Field(default=None, max_length=160)
    gateway_payment_intent: OptionalSanitizedStr = Field(default=None, max_length=160)
    checkout_url: str | None = Field(default=None, max_length=500)
    is_anonymous: bool = False


class DonationCreate(DonationBase):
    @model_validator(mode="after")
    def validate_phone_for_create(self) -> DonationCreate:
        self.donor_phone = _require_phone_unless_anonymous(
            is_anonymous=self.is_anonymous,
            donor_phone=self.donor_phone,
        )
        return self


class PublicDonationCreate(DonorBase):
    """Public donate surface — cannot set payment_status or gateway fields."""

    amount: float = Field(..., gt=0)
    currency: str = Field(default="INR", min_length=3, max_length=3)
    purpose: DonationPurpose = DonationPurpose.GENERAL
    payment_method: DonationPaymentMethod = DonationPaymentMethod.UPI
    transaction_reference: OptionalSanitizedStr = Field(default=None, max_length=120)
    donation_date: date = Field(default_factory=date.today)
    notes: OptionalFreeTextSanitizedStr = Field(default=None, max_length=500)
    is_anonymous: bool = False

    @model_validator(mode="after")
    def validate_public_rules(self) -> PublicDonationCreate:
        self.donor_phone = _require_phone_unless_anonymous(
            is_anonymous=self.is_anonymous,
            donor_phone=self.donor_phone,
        )
        if self.payment_method in UTR_REQUIRED_METHODS:
            if not _normalize_reference(self.transaction_reference):
                raise ValueError(
                    "transaction_reference (UTR) is required for "
                    f"{self.payment_method.value} payments"
                )
        return self

    def to_donation_create(self) -> DonationCreate:
        return DonationCreate(
            **self.model_dump(),
            payment_status=DonationPaymentStatus.PENDING,
            gateway_provider=None,
            gateway_session_id=None,
            gateway_payment_intent=None,
            checkout_url=None,
        )


class DonationUpdate(BaseModel):
    donor_name: SanitizedStr | None = Field(default=None, min_length=2, max_length=140)
    donor_email: str | None = Field(default=None, pattern=r"^[^@\s]+@[^@\s]+\.[^@\s]+$")
    donor_phone: OptionalSanitizedStr = Field(default=None, max_length=20)
    donor_address: OptionalSanitizedStr = Field(default=None, max_length=260)
    donor_pan: Annotated[
        str | None, Field(default=None, pattern=r"^[A-Z]{5}[0-9]{4}[A-Z]$")
    ] = None
    donor_type: DonorType | None = None
    amount: float | None = Field(default=None, gt=0)
    currency: str | None = Field(default=None, min_length=3, max_length=3)
    purpose: DonationPurpose | None = None
    payment_method: DonationPaymentMethod | None = None
    payment_status: DonationPaymentStatus | None = None
    transaction_reference: OptionalSanitizedStr = Field(default=None, max_length=120)
    donation_date: date | None = None
    notes: OptionalFreeTextSanitizedStr = Field(default=None, max_length=500)
    gateway_provider: OptionalSanitizedStr = Field(default=None, max_length=40)
    gateway_session_id: OptionalSanitizedStr = Field(default=None, max_length=160)
    gateway_payment_intent: OptionalSanitizedStr = Field(default=None, max_length=160)
    checkout_url: str | None = Field(default=None, max_length=500)
    is_anonymous: bool | None = None


class DonationMarkPaid(BaseModel):
    transaction_reference: OptionalSanitizedStr = Field(default=None, max_length=120)


class Donation(DonationBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    # ORM column is NOT NULL; empty string used when anonymous with no phone.
    donor_phone: str = ""
    receipt_number: str
    receipt_issued: bool = False
    receipt_issued_at: datetime | None = None
    receipt_snapshot: dict[str, Any] | None = None
    created_at: datetime
    updated_at: datetime


class DonationStatusEvent(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    donation_id: UUID
    from_status: str | None = None
    to_status: str
    transaction_reference: str | None = None
    actor_role: str | None = None
    actor_email: str | None = None
    note: str | None = None
    created_at: datetime


class DonationReceipt(BaseModel):
    receipt_number: str
    issued_at: datetime
    organization: str
    registration_note: str
    donation: Donation


class DonationCheckoutSession(BaseModel):
    donation_id: UUID
    provider: str
    checkout_url: str
    session_id: str
    publishable_key: str | None = None


class DonationListResponse(BaseModel):
    items: list[Donation]
    total: int
    paid: int
    pending: int
    failed: int
    refunded: int
    total_amount: float

File: frontend/lib/validation/donation.ts
import { z } from "zod";
import { idString, isoDateString, isoDateTimeString, nullableEmail, nullableString } from "./common";

export const donationPaymentMethodSchema = z.enum([
  "upi",
  "qr_code",
  "gpay",
  "cash",
  "cheque",
  "debit_card",
  "credit_card",
  "international_card",
  "stripe",
  "netbanking",
  "online_banking",
  "other",
]);

/** Public / CMS-seeded manual methods only. */
export const manualDonationPaymentMethodSchema = z.enum([
  "upi",
  "qr_code",
  "gpay",
  "cash",
  "cheque",
]);

export const utrRequiredPaymentMethods = new Set([
  "upi",
  "qr_code",
  "gpay",
  "cheque",
] as const);

export const donationPaymentStatusSchema = z.enum(["pending", "paid", "failed", "refunded"]);

export const donationPurposeSchema = z.enum([
  "general",
  "education",
  "healthcare",
  "livelihood",
  "relief",
  "environment",
]);

export const donorTypeSchema = z.enum([
  "individual",
  "corporate",
  "trust",
  "foundation",
  "other",
]);

export const donationSchema = z.object({
  id: idString,
  organization_id: idString,
  donor_name: z.string().min(1),
  donor_email: nullableEmail,
  donor_phone: z.string(),
  donor_address: nullableString,
  donor_pan: nullableString,
  donor_type: z.string().min(1),
  is_anonymous: z.boolean().default(false),
  amount: z.number().positive(),
  currency: z.string().min(3).max(3),
  purpose: donationPurposeSchema,
  payment_method: donationPaymentMethodSchema,
  payment_status: donationPaymentStatusSchema,
  transaction_reference: nullableString,
  donation_date: isoDateString,
  notes: nullableString,
  gateway_provider: nullableString,
  gateway_session_id: nullableString,
  gateway_payment_intent: nullableString,
  checkout_url: nullableString,
  receipt_number: z.string().min(1),
  receipt_issued: z.boolean(),
  receipt_issued_at: isoDateTimeString.nullable().optional().default(null),
  receipt_snapshot: z.record(z.string(), z.unknown()).nullable().optional().default(null),
  created_at: isoDateTimeString,
  updated_at: isoDateTimeString,
});

export const donationStatusEventSchema = z.object({
  id: idString,
  donation_id: idString,
  from_status: nullableString,
  to_status: z.string().min(1),
  transaction_reference: nullableString,
  actor_role: nullableString,
  actor_email: nullableString,
  note: nullableString,
  created_at: isoDateTimeString,
});

export const donationListResponseSchema = z.object({
  items: z.array(donationSchema),
  total: z.number().int().nonnegative(),
  paid: z.number().int().nonnegative(),
  pending: z.number().int().nonnegative(),
  failed: z.number().int().nonnegative(),
  refunded: z.number().int().nonnegative(),
  total_amount: z.number().nonnegative(),
});

export const donationCheckoutSessionSchema = z.object({
  donation_id: idString,
  provider: z.string().min(1),
  checkout_url: z.string().min(1),
  session_id: z.string().min(1),
  publishable_key: nullableString,
});

export const donationReceiptSchema = z.object({
  receipt_number: z.string().min(1),
  issued_at: isoDateTimeString,
  organization: z.string().min(1),
  registration_note: z.string(),
  donation: donationSchema,
});

export const donationFormSchema = z
  .object({
    donor_name: z.string().trim().min(2).max(140),
    donor_email: z.string().trim().email().or(z.literal("")),
    donor_phone: z.string().trim().max(20),
    donor_address: z.string().trim().max(260),
    donor_pan: z.string().trim().max(20),
    donor_type: donorTypeSchema.or(z.string().trim().max(40)),
    is_anonymous: z.boolean().default(false),
    amount: z.string().trim().min(1),
    currency: z.string().trim().min(3).max(3),
    purpose: donationPurposeSchema,
    payment_method: donationPaymentMethodSchema,
    payment_status: donationPaymentStatusSchema,
    transaction_reference: z.string().trim().max(120),
    donation_date: isoDateString,
    notes: z.string().trim().max(500),
  })
  .superRefine((values, ctx) => {
    if (!values.is_anonymous && values.donor_phone.length < 7) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ["donor_phone"],
        message: "Phone is required unless the donation is anonymous",
      });
    }
    if (values.is_anonymous && values.donor_phone && values.donor_phone.length < 7) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ["donor_phone"],
        message: "Phone must be at least 7 characters when provided",
      });
    }
  });

/** Public donate form: UTR required for UPI / QR / GPay / cheque. */
export const publicDonationFormSchema = donationFormSchema.superRefine((values, ctx) => {
  const method = values.payment_method;
  if (
    (method === "upi" || method === "qr_code" || method === "gpay" || method === "cheque") &&
    !values.transaction_reference.trim()
  ) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      path: ["transaction_reference"],
      message: "UTR / transaction reference is required for this payment method",
    });
  }
});

export type DonationParsed = z.infer<typeof donationSchema>;
export type DonationListResponseParsed = z.infer<typeof donationListResponseSchema>;
